### PR TITLE
updated pagination to comply with new standards

### PR DESCRIPTION
**Pull request by andre-urbani**
### What

Updated pagination to comply with the new API standards - https://github.com/ONSdigital/dp/blob/master/standards/API_STANDARDS.md 

- A limit of 0 now returns an empty list of items instead of all items
- The default limit has been set to 20
- Maximum limit has been set to 1000
- Negative values for offset and limit now result in an error
- Swagger spec and readme have been updated

### How to review

Confirm changes are correct

### Who can review

Anyone

### CODE REVIEW

File: swagger.yaml [hunk 1]
```diff
@@ -34,13 +34,28 @@ parameters:
     in: path
     type: string
     required: true
+  limit:
+    name: limit
+    description: "Maximum number of items that will be returned. A value of zero will return zero items. The default value is 20."
+    in: query
+    required: false
+    type: integer
```
Comment (sandypadmanabhan) [inline swagger.yaml:42]: can you add default: 20

File: swagger.yaml [hunk 2]
```diff
@@ -34,13 +34,28 @@ parameters:
...
+  offset:
+    name: offset
+    description: "Starting index of the items array that will be returned. By default it is zero, meaning that the returned items will start from the beginning."
+    in: query
+    required: false
+    type: integer
```
Comment (sandypadmanabhan) [inline swagger.yaml:48]: same here . We could add `default: 0`

File: api/codelists.go [hunk 3]
```diff
@@ -15,22 +15,38 @@ import (
...
+		if err != nil {
+			handleError(ctx, "failed to obtain a positive integer value for offset query parameter", log.Data{}, err, w)
+			return
+		}
 	}
 
-	log.Event(ctx, "attempting to get code lists", log.INFO, log.Data{"type": filterBy})
+	if limit > c.maxLimit {
+		logData["max_limit"] = c.maxLimit
+		handleError(ctx, "limit is greater than the maximum allowed", log.Data{}, err, w)
```
Comment (sandypadmanabhan) [inline api/codelists.go:47]: Not sure what this err would register as? Should it be something like err = errors.something?

File: api/codes.go [hunk 4]
```diff
@@ -18,20 +18,36 @@ func (c *CodeListAPI) getCodes(w http.ResponseWriter, r *http.Request) {
...
+		limit, err = ValidatePositiveInt(limitParameter)
+		if err != nil {
+			handleError(ctx, "failed to obtain a positive integer value for offset query parameter", log.Data{}, err, w)
+			return
+		}
+	}
+
+	if limit > c.maxLimit {
+		logData["max_limit"] = c.maxLimit
+		handleError(ctx, "limit is greater than the maximum allowed", log.Data{}, err, w)
```
Comment (sandypadmanabhan) [inline api/codes.go:50]: Same comment as mentioned above.

File: api/datasets.go [hunk 5]
```diff
@@ -19,20 +19,35 @@ func (c *CodeListAPI) getCodeDatasets(w http.ResponseWriter, r *http.Request) {
...
+		limit, err = ValidatePositiveInt(limitParameter)
+		if err != nil {
+			handleError(ctx, "failed to obtain a positive integer value for offset query parameter", log.Data{}, err, w)
+			return
+		}
+	}
+
+	if limit > c.maxLimit {
+		logData["max_limit"] = c.maxLimit
+		handleError(ctx, "limit is greater than the maximum allowed", log.Data{}, err, w)
```
Comment (sandypadmanabhan) [inline api/datasets.go:50]: same comment as above

File: api/editions.go [hunk 6]
```diff
@@ -16,27 +16,42 @@ func (c *CodeListAPI) getEditions(w http.ResponseWriter, r *http.Request) {
...
+		}
 	}
 
-	// get offset from query parameters, or default value
-	offset, err := GetPositiveIntQueryParameter(r.URL.Query(), "offset", c.defaultOffset)
-	if err != nil {
-		handleError(ctx, "failed to obtain a positive integer value for offset query parameter", log.Data{"offset": offset}, err, w)
+	if limit > c.maxLimit {
+		logData["max_limit"] = c.maxLimit
+		handleError(ctx, "limit is greater than the maximum allowed", log.Data{}, err, w)
```
Comment (sandypadmanabhan) [inline api/editions.go:48]: same comment as above

File: api/codelists.go [hunk 7]
```diff
@@ -15,22 +15,38 @@ import (
...
-	if err != nil {
-		handleError(ctx, "failed to obtain a positive integer value for limit query parameter", log.Data{"limit": limit}, err, w)
-		return
+	log.Event(ctx, "attempting to get code lists", log.INFO, log.Data{"type": filterBy})
+
+	if offsetParameter != "" {
+		logData["offset"] = offsetParameter
+		offset, err = ValidatePositiveInt(offsetParameter)
+		if err != nil {
+			handleError(ctx, "failed to obtain a positive integer value for offset query parameter", log.Data{}, err, w)
```
Comment (sandypadmanabhan) [inline api/codelists.go:31]: handleError currently has options for ErrNotFound or internalServerError. Just wondering ..
In this case if it fails, are we ok to return an internal server error? or do we want http.StatusBadRequest?

File: swagger.yaml (hunk 1 above)
Comment (andre-urbani) [reply swagger.yaml:42]: Good idea

File: api/codelists.go (hunk 3 above)
Comment (andre-urbani) [reply api/codelists.go:47]: I believe this is the correct implementation. `err` is defined on line 23 as type error, then passed to the `handleError` function which returns an error based on the following:
```
if err == driver.ErrNotFound {
		http.Error(w, err.Error(), http.StatusNotFound)
	} else {
		http.Error(w, internalServerErr, http.StatusInternalServerError)
	}
```
So I don't think there is any need for an if statement here

File: api/codelists.go (hunk 7 above)
Comment (andre-urbani) [reply api/codelists.go:31]: I'm not too sure on this one to be honest. This PR is just an update of a PR of mine that was approved by Carl a few days ago - https://github.com/ONSdigital/dp-code-list-api/pull/104/files

The error handling hasn't changed and Carl didn't comment on it previously so I think this is probably ok

Comment (andre-urbani) [reply api/codelists.go:31]: Happy to look into it further though if you think otherwise

File: README.md [hunk 8]
```diff
@@ -36,6 +36,9 @@ The endpoint `/health` checks the connection to the database and returns one of:
 | GRACEFUL_SHUTDOWN_TIMEOUT    | 5s                                     | The graceful shutdown timeout in seconds
 | HEALTHCHECK_INTERVAL         | 30s                                    | Time between calls to healthchecks
 | HEALTHCHECK_CRITICAL_TIMEOUT | 90s                                    | Timeout to consider a failing healthcheck critical
+| DEFAULT_MAXIMUM_LIMIT        | 1000                                   | Default maximum limit for pagination
+| DEFAULT_LIMIT                | 20                                     | Default limit for pagination
+| DEFAULT_OFFSET               | 0                                      | Default offset for pagination
 
 ### License
```
Comment (sandypadmanabhan) [inline README.md:44]: As we are here can you also update the License year in Readme.md and License.md

File: api/codelists_test.go [hunk 9]
```diff
@@ -92,28 +84,31 @@ func TestGetCodeLists(t *testing.T) {
...
+		r := httptest.NewRequest("GET", fmt.Sprintf("%s/code-lists?offset=-1&limit=-2", codeListURL), nil)
 		w := httptest.NewRecorder()
 
-		mockDatastore := &storetest.DataStoreMock{
-			GetCodeListsFunc: func(ctx context.Context, filterBy string) (*dbmodels.CodeListResults, error) {
-				return &dbCodeListResults, nil
-			},
-		}
+		api := CreateCodeListAPI(mux.NewRouter(), &storetest.DataStoreMock{}, codeListURL, datasetURL, defaultOffset, defaultLimit, maxLimit)
+		api.router.ServeHTTP(w, r)
```
[the remark this answers is not in the document]
Comment (andre-urbani) [reply api/codelists_test.go:99]: I think this is probably ok as it is

[the remark this answers is not in the document]
Comment (sandypadmanabhan) [reply api/codelists_test.go:149]: sure. No problem

**Review by sandypadmanabhan — APPROVED** [review]
Looks good

**Pull request merged**